### PR TITLE
Add filtering property to PanoramaSkyMaterial and switch from static shader to shader cache

### LINKED ISSUE #55163: Vulkan: PanoramaSkyMaterial cannot have filtering disabled on its texture anymore
### Godot version

4.0.dev (84f59c60b)

### System information

Fedora 34, GeForce GTX 1080 (NVIDIA 470.74)

### Issue description

Now that textures no longer store filtering options, PanoramaSky no longer makes it possible to disable filtering on the texture. Disabling filtering is useful for pixel art games that need panorama skies. `filter_linear` is hardcoded here:

https://github.com/godotengine/godot/blob/4ea87f36cd23494bfbde800b436467d3d0f82b52/scene/resources/sky_material.cpp#L356

We should figure out a way to add a PanoramaSky property that can be used to toggle between `filter_nearest` and `filter_linear` (default).

Low-resolution HDRI panorama for testing (from [Poly Haven](https://polyhaven.com/a/park_parking)): [park_parking_1k.zip](https://github.com/godotengine/godot/files/7575467/park_parking_1k.zip)

### Steps to reproduce

- Add a WorldEnvironment with a PanoramaSkyMaterial.

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by brennengreen**
Added the filter to the shader code in order to be able to enable/disable via editor property. As well as some help from @clayjohn to update from a single static shader to a shader cache.

*Bugsquad edit: This closes https://github.com/godotengine/godot/issues/55163.*

### CODE REVIEW

File: scene/resources/sky_material.cpp [hunk 1]
```diff
@@ -343,22 +360,24 @@ void PanoramaSkyMaterial::cleanup_shader() {
...
 void PanoramaSkyMaterial::_update_shader() {
+
+	String filter_str;
+	if (filter) {
+		filter_str = "filter_linear";
+	} else {
+		filter_str = "filter_nearest";
+	}
 	shader_mutex.lock();
 	if (shader.is_null()) {
```
Comment (clayjohn) [inline scene/resources/sky_material.cpp:371]: You'll need to remove this check to ensure that it updates on subsequent calls

File: scene/resources/sky_material.cpp [hunk 2]
```diff
@@ -308,12 +308,23 @@ Ref<Texture2D> PanoramaSkyMaterial::get_panorama() const {
 	return panorama;
 }
 
+void PanoramaSkyMaterial::set_filtering_enabled(bool p_enabled) {
+	filter = p_enabled;
+	RS::get_singleton()->material_set_param(_get_material(), "filter", filter);
```
Comment (clayjohn) [inline scene/resources/sky_material.cpp:313]: I don't see "filter" as a uniform in your shader, so this is likely updating nothing

File: scene/resources/sky_material.cpp [hunk 3]
```diff
@@ -308,12 +308,23 @@ Ref<Texture2D> PanoramaSkyMaterial::get_panorama() const {
...
+	return filter;
+}
+
 Shader::Mode PanoramaSkyMaterial::get_shader_mode() const {
 	return Shader::MODE_SKY;
 }
 
 RID PanoramaSkyMaterial::get_rid() const {
-	_update_shader();
+	((PanoramaSkyMaterial *)this)->_update_shader();
```
Comment (clayjohn) [inline scene/resources/sky_material.cpp:327]: I'm not sure this cast is needed

File: scene/resources/sky_material.cpp [hunk 4]
```diff
@@ -322,15 +333,21 @@ RID PanoramaSkyMaterial::get_rid() const {
 }
 
 RID PanoramaSkyMaterial::get_shader_rid() const {
-	_update_shader();
+	MutexLock lock(shader_mutex);
+	((PanoramaSkyMaterial *)this)->_update_shader();
```
Comment (clayjohn) [inline scene/resources/sky_material.cpp:337]: Mutex locks should have a corresponding unlock (see update shader below)

As above, I am not sure this cast is needed

**Review by clayjohn — COMMENTED** [review]
You're off to a great start! Good work so far.

**Comment by brennengreen** [comment]
@clayjohn 

When I remove the cast it gets upset about there not being any compatible type qualifiers  (I'm assuming this is because of `const`)

I pushed an updated commit with your suggestion, on my end though it still is not properly updating the shader (at all).

Glad to know `RS::get_singleton()->material_set_param(_get_material(), "filter", filter);` updates uniforms though, that was a point of confusion I was having. Seems like that's then unnecessary for this PR because we don't need a uniform for the filter just need to update it in the sample2d.

**Comment by clayjohn** [comment]
@brennengreen I came back to review this (sorry for the delay) and ended up just making a PR to your repo with the changes https://github.com/brennengreen/godot/pull/1

You can merge that to your branch and it will automatically update here!

**Comment by brennengreen** [comment]
@clayjohn 

There we go, should I rebase and squad to a single commit now?

**Comment by clayjohn** [comment]
> There we go, should I rebase and squad to a single commit now?

Yes, at the same time, you will also want to add the documentation line for ``filter``. You can do so by running the engine from the root folder of your godot directory and specifying ``--doctool .`` (explained in more detail [here](https://docs.godotengine.org/en/3.4/community/contributing/updating_the_class_reference.html#updating-the-documentation-template)).

File: doc/classes/PanoramaSkyMaterial.xml [hunk 5]
```diff
@@ -11,6 +11,8 @@
 	<tutorials>
 	</tutorials>
 	<members>
+		<member name="filter" type="bool" setter="set_filtering_enabled" getter="is_filtering_enabled" default="true">
+		</member>
```
Comment (clayjohn) [inline doc/classes/PanoramaSkyMaterial.xml:15]: You will need to fill in the docs to get CI to pass!

File: scene/resources/sky_material.cpp [hunk 6]
```diff
@@ -5,8 +5,8 @@
 /*                           GODOT ENGINE                                */
 /*                      https://godotengine.org                          */
 /*************************************************************************/
-/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
-/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/* Copyright (c) 2007-2021 Juan Linietsky, Ariel Manzur.                 */
```
Comment (clayjohn) [inline scene/resources/sky_material.cpp:8]: Looks like this change was accidentally made in your last update. Should be 2022 rather than 2021 here

**Comment by brennengreen** [comment]
@clayjohn Looks like we got all that cleared up, thank you a ton for the help :)

**Review by clayjohn — APPROVED** [review]

**Comment by clayjohn** [comment]
Great work! Thank you for your patience and perseverance!

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**